lib/livebook/completion.ex

```diff
@@ -0,0 +1,810 @@
+
+      {:dot_arity, path, hint} ->
+        complete_dot(path, List.to_string(hint), ctx)
+
+      {:dot_call, _path, _hint} ->
+        complete_default(ctx)
+
+      :expr ->
+        complete_default(ctx)
+
+      {:local_or_var, local_or_var} ->
+        complete_local_or_var(List.to_string(local_or_var), ctx)
+
+      {:local_arity, local} ->
+        complete_local(List.to_string(local), ctx)
+
+      {:local_call, _local} ->
+        complete_default(ctx)
+
+      # {:module_attribute, charlist}
+      # :none
+      _ ->
+        []
+    end
+  end
+
+  ## Complete dot
+
+  defp complete_dot(path, hint, ctx) do
+    case expand_dot_path(path, ctx) do
+      {:ok, mod} when is_atom(mod) and hint == "" ->
+        complete_module_member(mod, hint) ++ complete_module(mod, hint)
+
+      {:ok, mod} when is_atom(mod) ->
+        complete_module_member(mod, hint)
+
+      {:ok, map} when is_map(map) ->
+        complete_map_field(map, hint)
+
+      _ ->
+        []
+    end
+  end
+
+  defp expand_dot_path({:var, var}, ctx) do
+    Keyword.fetch(ctx.binding, List.to_atom(var))
+  end
+
+  defp expand_dot_path({:alias, alias}, ctx) do
+    {:ok, expand_alias(List.to_string(alias), ctx)}
+  end
+
+  defp expand_dot_path({:unquoted_atom, var}, _ctx) do
+    {:ok, List.to_atom(var)}
+  end
+
+  defp expand_dot_path({:dot, parent, call}, ctx) do
+    case expand_dot_path(parent, ctx) do
+      {:ok, %{} = map} -> Map.fetch(map, List.to_atom(call))
+      _ -> :error
+    end
+  end
+
+  defp complete_default(ctx) do
+    complete_local_or_var("", ctx)
+  end
+
+  defp complete_alias(hint, ctx) do
+    case split_at_last_occurrence(hint, ".") do
+      {hint, ""} ->
+        complete_elixir_root_module(hint) ++ complete_env_alias(hint, ctx)
+
+      {alias, hint} ->
+        mod = expand_alias(alias, ctx)
+        complete_module(mod, hint)
+    end
+  end
+
+  defp complete_module_member(mod, hint) do
+    complete_module_function(mod, hint) ++ complete_module_type(mod, hint)
+  end
+
+  defp complete_local_or_var(hint, ctx) do
+    complete_local(hint, ctx) ++ complete_variable(hint, ctx)
+  end
+
+  defp complete_local(hint, ctx) do
+    imports =
+      ctx.env
+      |> imports_from_env()
+      |> Enum.flat_map(fn {mod, funs} ->
+        complete_module_function(mod, hint, funs)
+      end)
+
+    special_forms = complete_module_function(Kernel.SpecialForms, hint)
+
+    imports ++ special_forms
+  end
+
+  defp complete_variable(hint, ctx) do
+    complete_key_value(ctx.binding, hint)
+  end
+
+  defp complete_map_field(map, hint) do
+    # Note: we need Map.to_list/1 in case this is a struct
+    complete_key_value(Map.to_list(map), hint)
+  end
+
+  defp complete_key_value(list, hint) do
+    for {key, value} <- list,
+        is_atom(key),
+        name = Atom.to_string(key),
+        String.starts_with?(name, hint),
+        do: %{
+          label: name,
+          kind: :variable,
+          detail: "variable",
```

Comment:
Is variable correct here? Field access/property/similar would be clearer, no? :)

Reply:
Property sounds more appropriate yup!